### DISCUSSION TITLE
How to make sure only chainlink automation calls my lottery contract and nobody else?

### DISCUSSION THREAD

**Discussion by Madhuvarun** [Cyfrin/foundry-full-course-cu#2269] [Q&A]
Hello,

I am currently watching foundry smart contract lottery in the foundry fundamentals module. In that module we are developing a smart contract to allow chainlink to automatically pick a winner. How to make sure that any EOA doesn't call performUpkeep function?

In the contract code we have the following functions.

```
function checkUpkeep(bytes memory /* checkData */ )
        public
        view
        override
        returns (bool upkeepNeeded, bytes memory /* performData */ )
    {
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool timePassed = ((block.timestamp - s_lastTimeStamp) > i_interval);
        bool hasPlayers = s_players.length > 0;
        bool hasBalance = address(this).balance > 0;
        upkeepNeeded = (timePassed && isOpen && hasBalance && hasPlayers);
        return (upkeepNeeded, "0x0"); // can we comment this out?
    }
```

```
function performUpkeep(bytes calldata /* performData */ ) external override {
        (bool upkeepNeeded,) = checkUpkeep("");
        // require(upkeepNeeded, "Upkeep not needed");
        if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(address(this).balance, s_players.length, uint256(s_raffleState));
        }

        s_raffleState = RaffleState.CALCULATING;

        // Will revert if subscription is not set and funded.
        uint256 requestId = s_vrfCoordinator.requestRandomWords(
            VRFV2PlusClient.RandomWordsRequest({
                keyHash: i_gasLane,
                subId: i_subscriptionId,
                requestConfirmations: REQUEST_CONFIRMATIONS,
                callbackGasLimit: i_callbackGasLimit,
                numWords: NUM_WORDS,
                extraArgs: VRFV2PlusClient._argsToBytes(
                    // Set nativePayment to true to pay for VRF requests with Sepolia ETH instead of LINK
                    VRFV2PlusClient.ExtraArgsV1({nativePayment: false})
                )
            })
        );
        // Quiz... is this redundant?
        emit RequestedRaffleWinner(requestId);
    }
```

checkUpkeep function is called periodically by chainlink automation to check if it returns true. If it returns true performUpkeep is called to pick a winner. Here anybody can call performUpkeep function when the time is right. This contract may not be vulnerable.I just want to make the smart contract be able to run only when chainlink calls it. How do I do this? I think adding a require in checkUpkeep function to check if msg.sender is chainlink automation contract. Does chainlink have a specific contract address to check for this?

Thank you

**Comment by EngrPips** [comment]
Hey @Madhuvarun, I understand your concern, but Patrick did not touch on how to restrict the Lottery contract `performUpkeep` function to only chainlink. Maybe that is because chainlink automation is a system of decentralized nodes, and you may not be sure what specific node will call the `performUpkeep` function. I don't think achieving this is impossible, though, but you might just need to do some personal research to find out how the chainlink automation node works. If they have maybe an interface in common that you can make a require statement against to safeguard the `performUpkeep` function from any other `msg.sender` that is not a chainlink node.

**Reply by Madhuvarun** [reply]
Thank you for your reply. I will research more on this.

**Reply by EngrPips** [reply]
Yeah, Take your time because it takes more time than you might expect when it is not a tutorial.
